### PR TITLE
Improved entity reference load time by caching swagger response in local storage

### PR DESCRIPTION
**Pull request by blackbaud-conorwright**
Improved entity reference load time by caching swagger response in local storage
Added angular local storage npm package

### CODE REVIEW

File: static/assets/js/entity-reference/app.1.2.js [hunk 1]
```diff
@@ -27,24 +27,39 @@
...
+            else {
+              handleSwaggerResponse.bind(this)(swaggerResponseCache.response);
+              bbWait.endPageWait();
+            }
         }
 
         function handleSuccess(response) {
+            localStorageService.set('swaggerResponseCache', {
+              'response': response,
+              'lastUpdatedDate': new Date()
```
Comment (Blackbaud-JasonTremper) [inline static/assets/js/entity-reference/app.1.2.js]: Might be easier to have "expirationDate" = new Date() + 12hours & then above check if now > expirationDate in onInit

File: static/assets/js/entity-reference/app.1.2.js [hunk 2]
```diff
@@ -27,24 +27,39 @@
         function onInit() {
             bbWait.beginPageWait({});
 
-            $http.get(this.swaggerUrl)
-                .then(handleSuccess.bind(this), handleError.bind(this))
-                .finally(function() { bbWait.endPageWait(); });
+            var swaggerResponseCache = localStorageService.get('swaggerResponseCache');
```
Comment (Blackbaud-JasonTremper) [inline static/assets/js/entity-reference/app.1.2.js]: Do you need to check localStorageService.isSupported first?

File: static/assets/js/entity-reference/app.1.2.js [hunk 3]
```diff
@@ -27,24 +27,39 @@
...
 
         function handleSuccess(response) {
+            localStorageService.set('swaggerResponseCache', {
+              'response': response,
+              'lastUpdatedDate': new Date()
+            });
+            handleSwaggerResponse.bind(this)(response);
+        }
+
+        function handleSwaggerResponse(response) {
```
Comment (Blackbaud-JasonTremper) [inline static/assets/js/entity-reference/app.1.2.js]: If this was just the swagger string, then you wouldn't need to cache the entire response in local storage...

File: static/assets/js/entity-reference/app.1.2.js [hunk 4]
```diff
@@ -27,24 +27,39 @@
         function onInit() {
             bbWait.beginPageWait({});
 
-            $http.get(this.swaggerUrl)
-                .then(handleSuccess.bind(this), handleError.bind(this))
-                .finally(function() { bbWait.endPageWait(); });
+            var swaggerResponseCache = localStorageService.get('swaggerResponseCache');
+            if (!swaggerResponseCache || Math.abs(new Date(swaggerResponseCache.lastUpdatedDate).getTime() - new Date().getTime()) / 36e5 >= 12) {
```
Comment (Blackbaud-JasonTremper) [inline static/assets/js/entity-reference/app.1.2.js]: See below comment for a possible easier implementation...

File: static/assets/js/entity-reference/app.1.2.js (hunk 1 above)
Comment (blackbaud-conorwright) [inline static/assets/js/entity-reference/app.1.2.js]: addressed. They have also been switched to milliseconds like we talked about :)

File: static/assets/js/entity-reference/app.1.2.js (hunk 3 above)
Comment (blackbaud-conorwright) [inline static/assets/js/entity-reference/app.1.2.js]: switched to just be the data. good catch 🐼

**Comment by Blackbaud-JasonTremper** [comment]
@Blackbaud-SteveBrush - I said the same thing offline. Angular 1 suggests using `this` over `$scope`, but I'm thinking in a future we should change this to use $scope so we don't have to use `bind()`

**Comment by Blackbaud-JasonTremper** [comment]
@blackbaud-conorwright - Did you mean to push the removed delay?

**Comment by blackbaud-conorwright** [comment]
I removed the delay when I figured it out. I was going to tell you @Blackbaud-JasonTremper , but I forgot to check back after your phone call.

@Blackbaud-SteveBrush , would you prefer that I change these methods to either:
1. Take in a "self" parameter to represent `this`
2. .bind(this) to the methods themselves, so that calls to them do not need to do it

**Pull request merged**